### DISCUSSION TITLE
API Endpoints aren't safe even with JWT Tokens.

### DISCUSSION THREAD

**Discussion by wulfharth7** [nextauthjs/next-auth#6298] [Help] [answered]
### Description of the Problem
I have lots of api routes, therefore lots of endpoints. The thing is these endpoints **_aren't safe despite users having roles_**. They can **see access token** from the network section e.g. from google chrome. I am using localhost for testing.
### Question
They'll be _able to use_ that on [postman and access](https://prnt.sc/DR0gN2_lZWEp) to all of the data. **How can I prevent that?**
### The Code
This is a regular session response I'll be getting back from my [...nextauth.js] endpoint
```
{
    "user": {
        "name": "x",
        "email": "x@gmail.com",
        "image": "https://lh3.googleusercontent.com/a/xxxxxx",
        "id": "42812490-3a84-4f95-b105-xxxxuuid",
        "role": "casual"
    },
    "expires": "2023-02-04T17:19:05.135Z",
    "accessToken": "xxxxxx-7eyekuKasCvtYXn13pFx4KuFDMFG83XzROdglWNgh9wyX6foRKu9KpbvUCFfTiegnQDzOwPruZ62EuFfoupHZ2VkpAaCgYKAVMSARISFQHUCsbC3VRVJ4pSZTtpl58iF0F5Jg0165"
}
```
this is my [...nextauth.js] file
```
import NextAuth from "next-auth"
import GoogleProvider from "next-auth/providers/google"
import GitHubProvider from 'next-auth/providers/github';
import { PrismaAdapter } from "@next-auth/prisma-adapter"
import prisma from "../../../lib/prismadb"
import { v4 } from "uuid";
import jwt from "jsonwebtoken"

export default NextAuth({
  adapter: PrismaAdapter(prisma),
  //debug: true,
  secret: process.env.NEXTAUTH_SECRET,
  
  providers: [
    // OAuth authentication providers
    GitHubProvider({
      clientId: process.env.GITHUB_ID,
      clientSecret: process.env.GITHUB_SECRET,
      checks:"both"
    }),
    GoogleProvider({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
      checks:"both"
    }),
  ],
  session:{
    strategy: 'jwt'
  },
  jwt: {
    maxAge: 60 * 60 * 24 * 30,
    async encode({ secret, token }) {
      return jwt.sign(token, secret)
    },
    async decode({ secret, token }) {
      return jwt.verify(token, secret)
    },
  },
  callbacks: {
    async signIn({ user, account }) {
        user.Type = "casual";
        user.password = user.id;
        user.certificateLink = "null";
        user.userID = v4();
        account.id = user.userID;
        account.role = user.Type
        return true
      },
    async redirect({ url, baseUrl }) {
      return baseUrl
    },
    async jwt({ token, user, account, profile, isNewUser}) {
      if (account){
        token.accessToken = account.access_token
        token.id = account.id
        token.role = account.role
      }
      return token
    },
    async session({session, user, token}) {
      session.accessToken = token.accessToken
      session.user.id = token.id
      session.user.role = token.role
      
      return session
    },

  },
  events: {
    async signIn(message) { /* on successful sign in */ },
    async signOut(message) { /* on signout */ },
    async createUser(message) { /* user created */ },
    async updateUser(message) { /* user updated - e.g. their email was verified */ },
    async linkAccount(message) { /* account (e.g. Twitter) linked to a user */ },
    async session(message) { /* session is active */ },
  },
  pages: {
    signIn: '/auth/signin',
    error: '/auth/error', // Error code passed in query string as ?error=
    verifyRequest: '/auth/verify-request', // (used for check email message)
  }

})
```
GET api/user endpoint
```
import prisma from '../../../lib/prisma'
import { v4 } from "uuid";
import { getToken } from "next-auth/jwt"
const secret = process.env.NEXTAUTH_SECRET


export default async function handle(req, res) {
  const token = await getToken({ req, secret, raw:true })
  console.log(token)
  if (token) {
    switch (req.method) {
      case 'POST':
        return createOnlyUser()
  
      case 'GET':
        return getAllUsers()
  
      default:
        res.json({ message: error });
      
      }
  } else {
    res.status(401)
  }
  res.end()
  

    async function getAllUsers(){
      const result = await prisma.user.findMany()
      res.json(result)
    }

    async function createOnlyUser(){
      const result = await prisma.user.create({
        data: {
          name: req.body.name,
          userID: v4(),
          password: req.body.password,
          email: req.body.email,
          certificateLink: req.body.certificateLink,
          Type: req.body.Type
        },
      })
      res.json(result)
    }
  }
```
token.role here doesn't work because it is actually an accessToken.

**Comment by programm-ingovals** [comment, marked as answer]
I'm not a NextAuth specialist or anything, but I fail to see the problem here. The holder of the access_token has access, that is the whole point of it.

Does it matter if they gain that data through your app or through postman? Are they accessing data in Postman they shouldn't have access to? Then the problem is with the access_token. Can you explain a bit better what the issue is?

If you are hoping that the access_token will just be stored server side and never enter the clientside you probably have to use a database for session management https://next-auth.js.org/faq#databases

I'm not promising that works, hopefully someone knowledgeable about NextAuth comes along to answer that.

**Reply by programm-ingovals** [reply]
> token.role here doesn't work because it is actually an accessToken.

Hang on, I think we aren't totally on the same page here.
`getToken` should give you the NextAuth token, so if you remove the `raw:true` option you should have an object the has the property role, so you could check that.

`token.accessToken` would give you the access token you added to the user, but I'm not certain that you should even bother with that. Is that an access_token you are getting from github/google ?

**Reply by programm-ingovals** [reply]
Also it seems that if you don't want the access token in the client side you shouldn't add it to the session here
```
async session({session, user, token}) {
session.accessToken = token.accessToken   <---- HERE!
session.user.id = token.id
session.user.role = token.role      
return session
},
```

If you aren't accessing any external API and just using the nextjs API endpoints I don't think you should bother with the access_token at all. The nextauth jwt token should be enough IMO. If you are dealing with roles you probably should use a database strategy to persist and ensure the roles of the users.

**Reply by wulfharth7** [reply]
the actual issue was they could reach to all api endpoints by using accessTokens. Even other users' datas. 

I've done more research. I was sending them the token in the session call back. So in the backend I use get server-side session instead of getToken. That solves issues.

I may try database strategy, I should take a look at it.

Thanks

**Reply by programm-ingovals** [reply]
I'm pretty sure the the auth is only checking if the access_token is valid, if you have extra requirements the onus is on you to ensure them. The thing is you can trust the access_token and the info in it. So a check, like `access_token.sub === data.owner` should do the trick.

**Reply by wulfharth7** [reply]
> I'm pretty sure the the auth is only checking if the access_token is valid, if you have extra requirements the onus is on you to ensure them. The thing is you can trust the access_token and the info in it. So a check, like `access_token.sub === data.owner` should do the trick.

https://next-auth.js.org/tutorials/securing-pages-and-api-routes
Yep, I was only checking if the token is valid or not the same way docs handle it. thanks.
